### PR TITLE
statsd agent on typescript

### PR DESCRIPTION
**Pull request by anton-sidelnikov**
Basic functionality

### CODE REVIEW

**Review by lego963 — APPROVED** [review]

File: source/app.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import { NetworkMonitor } from './monitors/network-monitor';
+import { DiskMonitor } from './monitors/disk-monitor';
+
+const config = loadCustomConfiguration()
+const monitors: any[] = [];
+
+function loadMonitors() {
+    const monitorNames = config.monitorNames;
+
+    for (let i = 0; i < monitorNames.length; i++) {
```
Comment (outcatcher) [inline source/app.ts]: Here and further: I suppose, it's better to use `monitorNames.forEach(name => ...)`

File: source/config.default.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+interface StatsdConfig {
+    prefix: string
+    env: string
+    host: string
+    debug: boolean
+}
+
+export function createDefaultConfig(): ConfigDefault {
```
Comment (outcatcher) [inline source/config.default.ts]: Maybe it makes sense to move the default configuration to json file and interfaces to `config.ts`

File: source/config.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        console.log('Loading configuration...')
+
+        customConfig = readFileSync(path +'/config.custom', 'utf8')
+        customConfig = JSON.parse(customConfig)
+    } else {
+        console.log('Creating default configuration file...')
+        const config = createDefaultConfig()
+        writeFileSync(path +'/config.custom', JSON.stringify(config) , 'utf-8')
+
+        loadCustomConfiguration(path)
```
Comment (outcatcher) [inline source/config.ts]: The `customConfig` seems to be `null` in `else` casee

File: source/monitors/cpu-monitor.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    collect(): void {
+        const intervalCpuTimes = this.getIntervalCpuTimes();
+
+        if (intervalCpuTimes == null)
+            return;
+
+        const totalIntervalCpuTime =
+            intervalCpuTimes.user + intervalCpuTimes.nice + intervalCpuTimes.sys + intervalCpuTimes.idle + intervalCpuTimes.irq;
```
Comment (outcatcher) [inline source/monitors/cpu-monitor.ts]: You can split this to multiple lines, e.g.
```
        const totalIntervalCpuTime =
            + intervalCpuTimes.user
            + intervalCpuTimes.nice
            + intervalCpuTimes.sys
            + intervalCpuTimes.idle
            + intervalCpuTimes.irq
```

File: source/monitors/cpu-monitor.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        this.setStatistics([
+            ['user', ((intervalCpuTimes.user / totalIntervalCpuTime) * 100).toFixed(2)],
```
Comment (outcatcher) [inline source/monitors/cpu-monitor.ts]: Maybe create a function for this rounding?

File: source/monitors/cpu-monitor.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        const newCpuTimes = {
+            user: 0,
+            nice: 0,
+            sys: 0,
+            idle: 0,
+            irq: 0
+        };
+
+        for (let i = 0; i < cpusInfo.length; i++) {
+            const cpu = cpusInfo[i];
```
Comment (outcatcher) [inline source/monitors/cpu-monitor.ts]: You can use `forEach` here too

File: source/monitors/cpu-monitor.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+            newCpuTimes.user += cpu.times.user;
+            newCpuTimes.nice += cpu.times.nice;
+            newCpuTimes.sys += cpu.times.sys;
+            newCpuTimes.idle += cpu.times.idle;
+            newCpuTimes.irq += cpu.times.irq;
```
Comment (outcatcher) [inline source/monitors/cpu-monitor.ts:85]: Maybe create a class with `add` method? So all same fields with be summed (pay attention to field types)

File: source/monitors/disk-monitor.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                ['readsMerged', disksInfo[disk].readsMerged],
+                ['sectorsRead', disksInfo[disk].sectorsRead],
+                ['msReading', disksInfo[disk].msReading],
+                ['writesCompleted', disksInfo[disk].writesCompleted],
+                ['writesMerged', disksInfo[disk].writesMerged],
+                ['sectorsWritten', disksInfo[disk].sectorsWritten],
+                ['msWriting', disksInfo[disk].msWriting],
+                ['iosPending', disksInfo[disk].iosPending],
+                ['msIo', disksInfo[disk].msIo],
+                ['msWeightedIo', disksInfo[disk].msWeightedIo])
```
Comment (outcatcher) [inline source/monitors/disk-monitor.ts]: Maybe `statisticsList.push(Object.entries(disksInfo[disk]))` will  suffice?

File: source/monitors/disk-monitor.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            for (const stat of statisticsList)
+                diskStatisticsList.push([`${disk}_${stat[0]}`, stat[1]])
```
Comment (outcatcher) [inline source/monitors/disk-monitor.ts]: You can use `statisticsList.map` function here

File: source/monitors/entities/monitor.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    sendStatistics(): void {
+        debug('Sending statistics...')
+
+        const statistics = this.statistics
+
+        for (let i = 0; i < statistics.length; i++) {
+            const statistic = statistics[i]
+
+            statistic.send()
```
Comment (outcatcher) [inline source/monitors/entities/monitor.ts]: Same `forEach` comment

File: source/monitors/entities/stat.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        this.statsdName = statsdName.split('.').map((s: string) => snakeCase(s)).join('.')
+        this.value = value
+        this.client = client(config)
+    }
+
+    send(): boolean {
+        debugSt('Sending statistic %s = %d', this.statsdName, this.value)
+        try {
+            this.client.gauge(this.statsdName, this.value)
+        } catch (err) {
```
Comment (outcatcher) [inline source/monitors/entities/stat.ts:24]: Maybe add some logging here? At least, `console.debug`.

File: source/monitors/memory-monitor.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    async collect(): Promise<void> {
+        const data = await mem()
+        try {
+            this.setStatistics([
+                ['free', data.free],
+                ['total', data.total],
+                ['used', data.used],
+            ])
```
Comment (outcatcher) [inline source/monitors/memory-monitor.ts]: Will `Object.entities(data)` suffice?

File: source/monitors/network-monitor.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        try {
+            this.setStatistics([
+                ['rx_bytes', stats[0].rx_bytes],
+                ['rx_dropped', stats[0].rx_dropped],
+                ['tx_bytes', stats[0].tx_bytes],
+                ['tx_dropped', stats[0].tx_dropped],
+                ['rx_sec', stats[0].rx_sec],
+                ['tx_sec', stats[0].tx_sec],
+                ['ms', stats[0].ms],
+            ])
```
Comment (outcatcher) [inline source/monitors/network-monitor.ts]: Same comment about `Object.entities`

File: source/utils/statsd-client.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+        host: config.statsdConfig.host,
+        cacheDns: true,
+        telegraf: true,
+        errorHandler: (err: Error) => console.error(err.stack || err),
+        prefix: `${config.statsdConfig.prefix}.${config.statsdConfig.env}.system.${hostname()}_`,
+        globalTags: {}
+    })
+
+    if (statsdConfig.debug) {
+        StatsdClient.prototype.sendMessage =
```
Comment (outcatcher) [inline source/utils/statsd-client.ts:18]: Is it possible to replace the method of the `client`, not of `StatsdClient.prototype`?

**Review by outcatcher — CHANGES_REQUESTED** [review]

File: source/config.ts (hunk 3 above)
Comment (anton-sidelnikov) [reply source/config.ts]: it recursive

Comment (outcatcher) [reply source/config.ts]: So it will be executed, but no result assignment will be done, so the return will still be empty

Comment (anton-sidelnikov) [reply source/config.ts]: fixed

File: source/monitors/disk-monitor.ts (hunk 8 above)
Comment (anton-sidelnikov) [reply source/monitors/disk-monitor.ts]: we need here first elem as metric name

File: source/config.ts (hunk 3 above)
Comment (outcatcher) [reply source/config.ts]: There should be `return loadCustomConfiguration(path)`

File: source/monitors/disk-monitor.ts (hunk 8 above)
Comment (outcatcher) [reply source/monitors/disk-monitor.ts]: `Object.entries` converts object to array of arrays
e.g. `Object.entries({1: 2, 3: 4})` will be `[["1", 2], ["3", 4]]`

File: source/utils/statsd-client.ts (hunk 14 above)
Comment (anton-sidelnikov) [reply source/utils/statsd-client.ts:18]: didn't find it

**Review by lego963 — APPROVED** [review]

**Review by lego963 — APPROVED** [review]

**Review by lego963 — APPROVED** [review]

**Review by outcatcher — APPROVED** [review]

**Pull request merged**